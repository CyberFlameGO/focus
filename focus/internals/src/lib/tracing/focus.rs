// module to collect tracing setup and config for the focus app itself

use std::io;

use anyhow::Result;
use tracing::dispatcher::DefaultGuard;
use tracing_appender::non_blocking::WorkerGuard;
use tracing_error::ErrorLayer;
use tracing_subscriber::prelude::*;
use tracing_subscriber::{self, util::SubscriberInitExt, EnvFilter};

#[derive(Debug)]
pub enum GuardWrapper {
    WorkerGuard(WorkerGuard),
    DefaultGuard(DefaultGuard),
}

#[derive(Debug, Default)]
/// opaque struct for returning tracing WorkerGuard instances to main
pub struct Guard {
    _inner: Vec<GuardWrapper>,
}

impl From<WorkerGuard> for Guard {
    fn from(wg: WorkerGuard) -> Self {
        Guard {
            _inner: vec![GuardWrapper::WorkerGuard(wg)],
        }
    }
}

#[derive(Debug, Default)]
pub struct TracingOpts {
    pub is_tty: bool,
    pub nocolor_requested: bool,
}

pub fn init_tracing(opts: TracingOpts) -> Result<Guard> {
    let TracingOpts {
        is_tty,
        nocolor_requested,
    } = opts;
    let use_color = is_tty && !nocolor_requested;
    let (non_blocking, guard) = tracing_appender::non_blocking(io::stderr());

    tracing_subscriber::registry()
        .with(ErrorLayer::default())
        .with(EnvFilter::new(
            std::env::var("RUST_LOG").unwrap_or_else(|_| "info".to_string()),
        ))
        .with(
            tracing_subscriber::fmt::layer()
                .with_span_events(tracing_subscriber::fmt::format::FmtSpan::CLOSE)
                .with_target(false)
                .with_writer(non_blocking)
                .with_ansi(use_color),
        )
        .try_init()?;

    Ok(Guard::from(guard))
}
