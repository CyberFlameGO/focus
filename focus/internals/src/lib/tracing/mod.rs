pub mod chrome;
pub mod focus;
pub mod git_trace2;

use std::path::PathBuf;

pub use chrome::Trace;
pub use focus::{init_tracing, Guard, TracingOpts};

use anyhow::Result;

#[cfg(test)]
pub(in crate::tracing) mod testing {
    use std::path::{Path, PathBuf};

    use anyhow::{anyhow, Result};

    pub const FIXTURE_RELPATH: &str = "src/lib/tracing/fixtures";

    pub fn manifest_dir() -> Result<PathBuf> {
        Ok(PathBuf::from(
            std::env::var_os("CARGO_MANIFEST_DIR").ok_or(anyhow!("CARGO_MANIFEST_DIR not set"))?,
        ))
    }

    pub fn manifest_relative_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        Ok(manifest_dir()?.join(s.as_ref()))
    }

    pub fn fixture_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        manifest_relative_path(FIXTURE_RELPATH).map(|p| p.join(s.as_ref()))
    }
}

/// returns the default system specific log location
#[cfg(target_os = "macos")]
pub fn log_dir() -> Result<PathBuf> {
    match dirs::home_dir().map(|pb| pb.join("Library/Logs/focus")) {
        Some(path) => Ok(path),
        None => Err(anyhow::anyhow!("HOME not defined")),
    }
}

#[cfg(not(any(
    target_os = "windows",
    target_os = "macos",
    target_os = "ios",
    target_arch = "wasm32"
)))]
pub fn log_dir() -> Result<PathBuf> {
    dirs::home_dir()
        .ok_or(anyhow::anyhow!("HOME not defined"))
        .map(|pb| pb.join(".local/focus/log"))
}
