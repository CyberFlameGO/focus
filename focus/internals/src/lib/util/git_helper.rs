use std::{
    ffi::{OsStr, OsString},
    fmt::Display,
    path::PathBuf,
    sync::Arc,
};

use anyhow::{bail, Context, Result};
use std::path::Path;
use std::process::Command;

use crate::{
    app::App,
    util::sandbox_command::{SandboxCommand, SandboxCommandOutput},
};

pub fn git_binary() -> OsString {
    OsString::from("git")
}

pub fn git_command(description: String, app: Arc<App>) -> Result<(Command, SandboxCommand)> {
    SandboxCommand::new(description, git_binary(), app)
}

pub fn remote_add<P: AsRef<Path>>(
    repo_path: P,
    name: &str,
    url: &OsStr,
    app: Arc<App>,
) -> Result<()> {
    let description = format!("Adding remote {} ({})", &name, &url.to_string_lossy());
    let (mut cmd, scmd) = git_command(description, app)?;
    scmd.ensure_success_or_log(
        cmd.current_dir(repo_path)
            .arg("remote")
            .arg("add")
            .arg(name)
            .arg(url),
        SandboxCommandOutput::Stderr,
        "git remote add",
    )
    .map(|_| ())
}

pub fn write_config<P: AsRef<Path>>(
    repo_path: P,
    key: &str,
    val: &str,
    app: Arc<App>,
) -> Result<()> {
    let description = format!("Setting Git config {} {}", key, val);
    let (mut cmd, scmd) = git_command(description, app)?;
    scmd.ensure_success_or_log(
        cmd.current_dir(repo_path).arg("config").arg(key).arg(val),
        SandboxCommandOutput::Stderr,
        "git config (write)",
    )
    .map(|_| ())
}

pub fn read_config<P: AsRef<Path>>(
    repo_path: P,
    key: &str,
    app: Arc<App>,
) -> Result<Option<String>> {
    let description = format!("Reading Git config {}", key);
    if let Ok(result) =
        run_consuming_stdout(description, repo_path, vec!["config", key], app.clone())
    {
        return Ok(Some(result));
    }

    Ok(None)
}

pub fn unset_config<P: AsRef<Path>>(repo_path: P, key: &str, app: Arc<App>) -> Result<()> {
    let description = format!("git config --unset {}", key);
    let (mut cmd, _scmd) = git_command(description, app)?;
    cmd.arg("config")
        .arg("--unset")
        .arg(key)
        .current_dir(repo_path)
        .status()
        .with_context(|| format!("Running `git config --unset {}` failed", key))?;
    Ok(())
}

pub fn run_consuming_stdout<P, I, S>(
    description: String,
    repo: P,
    args: I,
    app: Arc<App>,
) -> Result<String>
where
    P: AsRef<Path>,
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    let (mut cmd, scmd) = git_command(description, app)?;
    if let Err(e) = cmd.current_dir(repo).args(args).status() {
        scmd.log(SandboxCommandOutput::Stderr, &"git command")?;
        bail!("git command failed: {}", e);
    }
    let mut stdout_contents = String::new();
    scmd.read_to_string(SandboxCommandOutput::Stdout, &mut stdout_contents)?;
    Ok(stdout_contents.trim().to_owned())
}

pub fn find_top_level(app: Arc<App>, path: &Path) -> Result<PathBuf> {
    if let Ok(path) = std::fs::canonicalize(path) {
        Ok(PathBuf::from(
            run_consuming_stdout(
                format!("Finding top level of repo in {}", path.display()),
                path,
                vec!["rev-parse", "--show-toplevel"],
                app,
            )
            .context("Finding the repo's top level failed")?,
        ))
    } else {
        bail!(
            "Could not canonicalize repository path '{}'",
            &path.display()
        );
    }
}

pub fn get_current_revision(app: Arc<App>, repo: &Path) -> Result<String> {
    run_consuming_stdout(
        format!("Determining the current commit in repo {}", repo.display()),
        repo,
        vec!["rev-parse", "HEAD"],
        app,
    )
}

pub fn get_current_branch(app: Arc<App>, repo: &Path) -> Result<String> {
    run_consuming_stdout(
        format!("Determining the current branch in repo {}", repo.display()),
        repo,
        vec!["branch", "--show-current"],
        app,
    )
}

// Switches to a branch in a given repository, switching back to the previous branch afterwards
pub struct BranchSwitch {
    app: Arc<App>,
    repo: PathBuf,
    refname: String,
    alternate: Option<PathBuf>,
    switch_back: Option<String>,
}

impl BranchSwitch {
    #[allow(unused)]
    pub fn permanent(
        app: Arc<App>,
        repo: PathBuf,
        refname: String,
        alternate: Option<PathBuf>,
    ) -> Result<Self> {
        let instance = Self {
            app,
            repo,
            refname,
            alternate,
            switch_back: None,
        };

        instance
            .switch(&instance.refname, true)
            .context("Switching branches failed")?;

        Ok(instance)
    }

    pub fn temporary(
        app: Arc<App>,
        repo: PathBuf,
        refname: String,
        alternate: Option<PathBuf>,
    ) -> Result<Self> {
        let current_branch = {
            let hint = get_current_branch(app.clone(), repo.as_path())?
                .trim()
                .to_owned();
            if hint.is_empty() {
                app.ui().log(
                    "Branch Switch",
                    format!(
                        "Couldn't determine the current branch in {}, using the default 'master'.",
                        repo.display()
                    ),
                );

                String::from("master")
            } else {
                hint
            }
        };

        let instance = Self {
            app,
            repo,
            refname,
            alternate,
            switch_back: Some(current_branch),
        };

        instance
            .switch(&instance.refname, true)
            .context("Switching branches failed")?;

        Ok(instance)
    }

    fn switch(&self, branch_name: &str, detach: bool) -> Result<()> {
        let attachment_description = if detach { "detached" } else { "attached" };
        let description = format!(
            "Switching to {} in {} ({})",
            &branch_name,
            self.repo.display(),
            attachment_description,
        );
        let (mut cmd, scmd) = git_command(description.clone(), self.app.clone())?;
        let cmd = cmd.arg("switch").current_dir(&self.repo);
        if detach {
            cmd.arg("--detach");
        }
        cmd.arg(&branch_name);

        if let Some(alternate_path) = &self.alternate {
            cmd.env(
                "GIT_ALTERNATE_OBJECT_DIRECTORIES",
                alternate_path.as_os_str(),
            );
        }
        scmd.ensure_success_or_log(cmd, SandboxCommandOutput::Stderr, &description)?;

        Ok(())
    }
}

impl Drop for BranchSwitch {
    fn drop(&mut self) {
        if let Some(refname_to_switch_back_to) = &self.switch_back {
            self.switch(&refname_to_switch_back_to, false)
                .expect("Switching back to the original branch failed");
        }
    }
}

/// The state of a repository encompassing its origin URL and current commit ID.
#[derive(Debug, Eq, PartialEq, Hash, Clone)]
pub struct RepoState {
    origin_url: String,
    commit_id: String,
}

impl RepoState {
    pub fn new(repo_path: &dyn AsRef<Path>, app: Arc<App>) -> Result<Self> {
        let origin_url = run_consuming_stdout(
            "Reading origin URL".to_owned(),
            repo_path,
            vec!["remote", "get-url", "origin"],
            app.clone(),
        )
        .context("Failed to determine the origin URL")?;

        let commit_id = run_consuming_stdout(
            "Determining commit ID".to_owned(),
            repo_path,
            vec!["rev-parse", "HEAD"],
            app.clone(),
        )
        .context("Failed to determine the commit ID")?;

        Ok(Self {
            origin_url,
            commit_id,
        })
    }
}

impl Display for RepoState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}?@={}", self.origin_url, self.commit_id)
    }
}
