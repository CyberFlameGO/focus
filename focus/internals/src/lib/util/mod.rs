pub mod backed_up_file;
pub mod git_helper;
pub mod lock_file;
pub mod periodic;
pub mod repo_paths;
pub mod sandbox;
pub mod sandbox_command;
pub mod temporary_working_directory;
