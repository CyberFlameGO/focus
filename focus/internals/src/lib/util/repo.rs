use std::{path::Path, str::FromStr, sync::Arc};

use anyhow::{bail, Context, Result};
use uuid::Uuid;

use crate::{app::App, ui::ProgressReporter, util::git_helper};

