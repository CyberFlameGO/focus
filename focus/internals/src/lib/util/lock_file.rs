use anyhow::{bail, Context, Result};
use std::fs::File;
use std::io::{BufWriter, Write};
use std::path::{Path, PathBuf};
use tracing::{error, warn};

pub struct LockFile {
    path: PathBuf,
    fd: i32,
}

impl LockFile {
    // Try to obtain an exclusively locked file at the given path. It must not exist.
    pub fn new(path: &Path) -> Result<Self> {
        use std::os::unix::prelude::*;

        if std::fs::metadata(path).is_ok() {
            error!("Another process is holding a lock on {}", path.display());
            error!(
                "The lock is held by {}",
                std::fs::read_to_string(path).context("Failed reading lockfile")?
            );
            bail!("Another process holds a lock");
        }

        match File::create(path) {
            Ok(file) => {
                Self::write_process_description(&file)?;
                let fd = file.into_raw_fd();

                if let Err(e) = Self::acqrel_lock(fd, true) {
                    bail!(
                        "Acquiring exclusive advisory lock on {} failed: {}",
                        path.display(),
                        e
                    );
                }
                Ok(Self {
                    path: path.to_owned(),
                    fd,
                })
            }
            Err(e) => {
                bail!("Creating lock file {} failed: {}", path.display(), e);
            }
        }
    }

    fn write_process_description(file: &File) -> Result<()> {
        let mut buffered_writer = BufWriter::new(file);
        writeln!(
            buffered_writer,
            "PID {} started by {} on host {}",
            std::process::id(),
            whoami::username(),
            whoami::hostname(),
        )?;
        buffered_writer.flush()?;
        file.sync_all()?;

        Ok(())
    }

    fn acqrel_lock(fd: i32, lock: bool) -> Result<()> {
        use nix::*;

        let mut fls: libc::flock = unsafe { core::mem::zeroed() };
        if lock {
            fls.l_type = libc::F_WRLCK as libc::c_short;
        } else {
            fls.l_type = libc::F_UNLCK as libc::c_short;
        }
        fls.l_whence = libc::SEEK_SET as libc::c_short;
        fls.l_start = 0;
        fls.l_len = 0;
        fls.l_pid = unistd::getpid().as_raw();

        if let Err(e) = fcntl::fcntl(fd, fcntl::FcntlArg::F_SETLKW(&fls)) {
            bail!("fnctl error: {}", e);
        }

        Ok(())
    }
}

impl Drop for LockFile {
    fn drop(&mut self) {
        if let Err(e) = Self::acqrel_lock(self.fd, false) {
            warn!(
                ?self.path,
                ?e,
                "Releasing advisory lock on file failed",
            );
        }
        if let Err(e) = std::fs::remove_file(self.path.as_path()) {
            warn!("Removing lock file {} failed: {}", self.path.display(), e);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use anyhow::Result;
    use std::cell::Cell;
    use tempfile::tempdir;

    #[test]
    fn creating_a_lock() -> Result<()> {
        let dir = tempdir()?;
        let path = dir.path().join("lockfile");

        LockFile::new(&path).expect("should have succeeded");

        Ok(())
    }

    #[test]
    fn failing_to_create_a_duplicate_lock() -> Result<()> {
        let dir = tempdir()?;
        let path = dir.path().join("lockfile");

        let _a = LockFile::new(&path).expect("should have succeeded");
        let _b = LockFile::new(&path).err().expect("should have failed");

        Ok(())
    }

    #[test]
    fn failing_to_create_a_lock_in_an_inextant_directory() -> Result<()> {
        let path: Cell<Option<PathBuf>> = Cell::new(None);
        {
            let dir = tempdir()?;
            let inner_path = dir.path().join("lockfile");
            path.replace(Some(inner_path));
        }

        let path = path.take().unwrap();
        LockFile::new(path.as_path())
            .err()
            .expect("should have failed");

        Ok(())
    }
}
