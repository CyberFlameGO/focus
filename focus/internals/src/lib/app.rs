use std::borrow::Borrow;
use std::sync::Arc;

use crate::{ui::UserInterface, util::sandbox::Sandbox};
use anyhow::{Context, Result};
use std::time::SystemTime;
use ti_library::tool_insights_client::ToolInsightsClient;

#[derive(Clone)]
pub struct App {
    ui: Arc<UserInterface>,
    sandbox: Arc<Sandbox>,
    tool_insights_client: ToolInsightsClient,
}

impl App {
    pub fn new(preserve_sandbox_contents: bool, interactive: bool) -> Result<Self> {
        let ui = Arc::from(UserInterface::new(interactive).context("Failed to start UI")?);
        let sandbox =
            Arc::from(Sandbox::new(preserve_sandbox_contents).context("Failed to create sandbox")?);
        let tool_insights_client = ToolInsightsClient::new(
            // TODO: get this from toml file
            "focus".to_string(),
            // TODO: get this from toml file
            "0.3".to_string(),
            SystemTime::now(),
        );
        Ok(Self {
            ui,
            sandbox,
            tool_insights_client,
        })
    }

    /// Get a reference to the app's ui.
    pub fn ui(&self) -> Arc<UserInterface> {
        self.ui.clone()
    }

    /// Get a reference to the app's sandbox.
    pub fn sandbox(&self) -> Arc<Sandbox> {
        self.sandbox.clone()
    }

    /// Get a reference to the app's tool-insights client.
    pub fn tool_insights_client(&self) -> &ToolInsightsClient {
        self.tool_insights_client.borrow()
    }
}
