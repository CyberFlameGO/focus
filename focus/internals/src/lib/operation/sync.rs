use crate::model::repo::Repo;
use crate::{index::RocksDBMemoizationCacheExt, locking};

use crate::operation::util::perform;
use crate::target::TargetSet;
use content_addressed_cache::RocksDBCache;
use focus_util::app::App;
use focus_util::backed_up_file::BackedUpFile;
use tracing::warn;

use std::convert::TryFrom;

use std::path::Path;

use std::sync::Arc;

use anyhow::{bail, Context, Result};

pub struct SyncResult {
    pub checked_out: bool,
    pub commit_id: Option<git2::Oid>,
    pub skipped: bool,
}

/// Synchronize the sparse repo's contents with the build graph. Returns whether a checkout actually occured.
pub fn run(sparse_repo: &Path, preemptive: bool, app: Arc<App>) -> Result<SyncResult> {
    let repo = Repo::open(sparse_repo, app.clone()).context("Failed to open the repo")?;

    let _lock = locking::hold_lock(sparse_repo, Path::new("sync.lock"))
        .context("Failed to obtain synchronization lock")?;

    let sparse_profile_path = repo.git_dir().join("info").join("sparse-checkout");
    if !sparse_profile_path.is_file() {
        bail!("This does not appear to be a focused repo -- it is missing a sparse checkout file");
    }

    if preemptive && !repo.get_preemptive_sync_enabled()? {
        return Ok(SyncResult {
            checked_out: false,
            commit_id: None,
            skipped: true,
        });
    }

    let selections = repo.selection_manager()?;
    let selection = selections.computed_selection()?;
    let targets = TargetSet::try_from(&selection).context("constructing target set")?;

    // Add target/project to TI data.
    let app_for_ti_client = app.clone();
    let ti_client = app_for_ti_client.tool_insights_client();

    let backed_up_sparse_profile: Option<BackedUpFile> = if preemptive {
        None
    } else {
        super::ensure_clean::run(sparse_repo, app.clone())
            .context("Failed trying to determine whether working trees were clean")?;

        ti_client
            .get_context()
            .add_to_custom_map("total_target_count", targets.len().to_string());
        ti_client.get_context().add_to_custom_map(
            "user_selected_project_count",
            selection.projects.len().to_string(),
        );
        ti_client.get_context().add_to_custom_map(
            "user_selected_target_count",
            selection.targets.len().to_string(),
        );

        Some(BackedUpFile::new(&sparse_profile_path)?)
    };

    let head_commit = repo.get_head_commit().context("Resolving head commit")?;

    // Figure out if this repo has a "master" branch or "main" branch.
    let primary_branch_name =
        primary_branch_name(&repo).context("Determining primary branch name")?;

    let commit = if preemptive {
        if let Some(prefetch_commit) = repo
            .get_prefetch_head_commit("origin", primary_branch_name.as_str())
            .context("Resolving prefetch head commit")?
        {
            prefetch_commit
        } else {
            bail!("No prefetch commit found for preemptive sync");
        }
    } else {
        head_commit
    };

    if preemptive {
        if let Some(working_tree) = repo.working_tree() {
            if let Ok(Some(sync_point)) = working_tree.read_sync_point_ref() {
                if sync_point == commit.id() {
                    // The sync point is already set to this ref. We don't need to bother.
                    warn!("Skipping preemptive synchronization because the commit to sync is the same as that of the sync point");
                    return Ok(SyncResult {
                        checked_out: false,
                        commit_id: Some(commit.id()),
                        skipped: true,
                    });
                }
            }
        }
    }

    let (pattern_count, checked_out) = perform("Computing the new sparse profile", || {
        let odb = RocksDBCache::new(repo.underlying());
        repo.sync(
            commit.id(),
            &targets,
            false,
            &repo.config().index,
            app.clone(),
            &odb,
        )
        .context("Sync failed")
    })?;

    if !preemptive {
        ti_client
            .get_context()
            .add_to_custom_map("pattern_count", pattern_count.to_string());
        perform("Updating the sync point", || {
            repo.working_tree().unwrap().write_sync_point_ref()
        })?;

        // The profile was successfully applied, so do not restore the backup.
        backed_up_sparse_profile.unwrap().set_restore(false);
    }

    Ok(SyncResult {
        checked_out,
        commit_id: Some(commit.id()),
        skipped: false,
    })
}

fn primary_branch_name(repo: &Repo) -> Result<String> {
    let underlying = repo.underlying();
    if underlying.find_reference("refs/heads/master").is_ok() {
        Ok(String::from("master"))
    } else if underlying.find_reference("refs/heads/main").is_ok() {
        Ok(String::from("main"))
    } else {
        bail!("Could not determine primary branch name")
    }
}

#[cfg(test)]
mod testing {
    use std::{collections::HashSet, path::Path};

    use anyhow::Result;
    use maplit::hashset;

    use focus_testing::{init_logging, ScratchGitRepo};
    use focus_util::app;

    use crate::{
        model::{repo::Repo, selection::OperationAction},
        operation::{
            self,
            testing::integration::{RepoDisposition, RepoPairFixture},
        },
    };

    fn add_updated_content(scratch_repo: &ScratchGitRepo) -> Result<git2::Oid> {
        // Commit new files affecting the build graph to the dense repo
        let build_bazel_content = r#"filegroup(
            name = "excerpts",
            srcs = [
                "catz.txt",
            ],
            visibility = [
                "//visibility:public",
            ],
        )"#;
        scratch_repo.write_and_commit_file(
            Path::new("x/BUILD.bazel"),
            build_bazel_content.as_bytes(),
            "Add excerpts",
        )?;
        let catz_txt_content = r#"The Naming of Cats is a difficult matter,
        It isn't just one of your holiday games
                )"#;
        scratch_repo.write_and_commit_file(
            Path::new("x/catz.txt"),
            catz_txt_content.as_bytes(),
            "Add excerpts",
        )
    }

    #[test]
    fn sync_upstream_changes() -> Result<()> {
        init_logging();

        let fixture = RepoPairFixture::new()?;

        fixture.perform_clone()?;
        let _ = add_updated_content(&fixture.dense_repo)?;

        // Fetch in the sparse repo from the dense repo
        fixture.perform_pull(RepoDisposition::Sparse, "origin", "main")?;

        // Make sure that the graph is seen as having changed
        assert_eq!(
            operation::detect_build_graph_changes::run(
                &fixture.sparse_repo_path,
                vec![],
                fixture.app.clone(),
            )?,
            app::ExitCode(1)
        );

        // Sync in the sparse repo
        let _ = operation::sync::run(&fixture.sparse_repo_path, false, fixture.app.clone())?;

        let x_dir = fixture.sparse_repo_path.join("x");
        assert!(!x_dir.is_dir());

        // Add as a target
        operation::selection::add(
            &fixture.sparse_repo_path,
            false,
            vec![String::from("bazel://x/...")],
            fixture.app.clone(),
        )?;

        // Sync
        let _ = operation::sync::run(&fixture.sparse_repo_path, false, fixture.app.clone())?;

        assert!(x_dir.is_dir());

        Ok(())
    }

    #[test]
    fn sync_layer_manipulation() -> Result<()> {
        init_logging();

        let fixture = RepoPairFixture::new()?;
        fixture.perform_clone()?;

        let selected_project_names = || -> Result<HashSet<String>> {
            Ok(fixture
                .sparse_repo()?
                .selection_manager()?
                .computed_selection()?
                .projects
                .iter()
                .filter_map(|project| {
                    if project.mandatory {
                        None
                    } else {
                        Some(project.name.to_owned())
                    }
                })
                .collect::<HashSet<String>>())
        };

        let project_a_label = String::from("team_banzai/project_a");
        let project_b_label = String::from("team_zissou/project_b");

        let path = fixture.sparse_repo_path.clone();
        let library_a_dir = path.join("library_a");
        let project_a_dir = path.join("project_a");
        let library_b_dir = path.join("library_b");
        let project_b_dir = path.join("project_b");
        let profile_path = path.join(".git").join("info").join("sparse-checkout");

        {
            let selected_names = selected_project_names()?;
            assert_eq!(selected_names, hashset! {});
        }
        insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);

        assert!(!library_b_dir.is_dir());
        assert!(!project_b_dir.is_dir());
        operation::selection::add(
            &path,
            false,
            vec![project_b_label.clone()],
            fixture.app.clone(),
        )?;
        {
            let selected_names = selected_project_names()?;
            assert_eq!(selected_names, hashset! { project_b_label.clone() })
        }
        operation::sync::run(&path, false, fixture.app.clone())?;

        insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);
        assert!(library_b_dir.is_dir());
        assert!(project_b_dir.is_dir());

        assert!(!library_a_dir.is_dir());
        assert!(!project_a_dir.is_dir());
        operation::selection::add(
            &path,
            false,
            vec![project_a_label.clone()],
            fixture.app.clone(),
        )?;
        {
            let selected_names = selected_project_names()?;
            assert_eq!(
                selected_names,
                hashset! { project_a_label.clone(), project_b_label.clone() }
            )
        }
        operation::sync::run(&path, false, fixture.app.clone())?;
        insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);
        assert!(library_a_dir.is_dir());
        assert!(project_a_dir.is_dir());

        operation::selection::remove(&path, false, vec![project_a_label], fixture.app.clone())?;
        {
            let selected_names = selected_project_names()?;
            assert_eq!(selected_names, hashset! { project_b_label.clone() })
        }
        operation::sync::run(&path, false, fixture.app.clone())?;
        insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);
        assert!(!library_a_dir.is_dir());
        assert!(!project_a_dir.is_dir());

        operation::selection::remove(&path, false, vec![project_b_label], fixture.app.clone())?;
        {
            let selected_names = selected_project_names()?;
            assert_eq!(selected_names, hashset! {});
        }
        operation::sync::run(&path, false, fixture.app.clone())?;
        insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);

        assert!(!library_b_dir.is_dir());
        assert!(!project_b_dir.is_dir());

        Ok(())
    }

    #[test]
    fn sync_adhoc_manipulation() -> Result<()> {
        init_logging();

        let fixture = RepoPairFixture::new()?;
        fixture.perform_clone()?;

        let path = fixture.sparse_repo_path.clone();
        let library_b_dir = path.join("library_b");
        let targets = vec![String::from("bazel://library_b/...")];
        let mut selections = fixture.sparse_repo()?.selection_manager()?;

        assert!(selections.mutate(OperationAction::Add, &targets)?);
        selections.save()?;
        operation::sync::run(&path, false, fixture.app.clone())?;
        assert!(library_b_dir.is_dir());

        // operation::adhoc::pop(fixture.sparse_repo_path.clone(), 1)?;
        assert!(selections.mutate(OperationAction::Remove, &targets)?);
        selections.save()?;
        operation::sync::run(&path, false, fixture.app.clone())?;
        assert!(!library_b_dir.is_dir());

        Ok(())
    }

    #[test]
    fn clone_contains_top_level() -> Result<()> {
        init_logging();

        let fixture = RepoPairFixture::new()?;
        fixture.perform_clone()?;

        let sparse_repo = fixture.sparse_repo()?;
        let outlining_tree = sparse_repo.outlining_tree().unwrap();
        let underlying = outlining_tree.underlying();
        let outlining_tree_root = underlying.work_dir();

        let top_level_bazelisk_rc = outlining_tree_root.join(".bazeliskrc");

        let outlining_tree_git_dir = sparse_repo
            .git_dir()
            .join("worktrees")
            .join("outlining-tree");

        let profile =
            std::fs::read_to_string(outlining_tree_git_dir.join("info").join("sparse-checkout"))?;
        insta::assert_snapshot!(&profile);

        assert!(top_level_bazelisk_rc.is_file());

        Ok(())
    }

    #[test]
    fn sync_skips_checkout_with_unchanged_profile() -> Result<()> {
        init_logging();

        let fixture = RepoPairFixture::new()?;
        fixture.perform_clone()?;

        let path = fixture.sparse_repo_path.clone();
        let targets = vec![String::from("bazel://library_b/...")];
        let mut selections = fixture.sparse_repo()?.selection_manager()?;

        assert!(selections.mutate(OperationAction::Add, &targets)?);
        selections.save()?;
        {
            let result = operation::sync::run(&path, false, fixture.app.clone())?;
            assert!(!result.skipped);
            assert!(result.checked_out);
        }

        // Subsequent sync does not perform a checkout.
        {
            let result = operation::sync::run(&path, false, fixture.app.clone())?;
            assert!(!result.skipped);
            assert!(!result.checked_out);
        }

        Ok(())
    }

    #[test]
    fn sync_configures_working_and_outlining_trees() -> Result<()> {
        init_logging();

        let fixture = RepoPairFixture::new()?;
        fixture.perform_clone()?;

        // Check working tree
        let working_tree_config = fixture.sparse_repo()?.underlying().config()?.snapshot()?;
        assert!(working_tree_config.get_bool("index.sparse")?);
        assert!(working_tree_config.get_bool("core.untrackedCache")?);

        // Check outlining tree
        let outlining_tree_config = fixture
            .sparse_repo()?
            .outlining_tree()
            .unwrap()
            .underlying()
            .git_repo()
            .config()?
            .snapshot()?;
        assert!(outlining_tree_config.get_bool("index.sparse")?);
        assert!(outlining_tree_config.get_bool("core.untrackedCache")?);

        Ok(())
    }

    struct PreemptiveSyncFixture {
        pub underlying: RepoPairFixture,
        pub repo: Repo,
        pub commit_id: git2::Oid,
    }

    impl PreemptiveSyncFixture {
        fn new() -> Result<Self> {
            let fixture = RepoPairFixture::new()?;

            fixture.perform_clone()?;
            add_updated_content(&fixture.dense_repo)?;

            let fetched_commits = fixture.perform_fetch(RepoDisposition::Sparse, "origin")?;
            assert_eq!(fetched_commits.len(), 1);
            let commit_id = fetched_commits[0];

            let repo = Repo::open(&fixture.sparse_repo_path, fixture.app.clone())?;
            repo.set_preemptive_sync_enabled(true)?;

            // Set the prefetch ref
            repo.underlying().reference(
                "refs/prefetch/remotes/origin/main",
                commit_id,
                true,
                "Emulated prefetch ref",
            )?;

            Ok(PreemptiveSyncFixture {
                underlying: fixture,
                repo,
                commit_id,
            })
        }
    }

    #[test]
    fn preemptive_sync() -> Result<()> {
        init_logging();

        let fixture = PreemptiveSyncFixture::new()?;

        // Sync preemptively
        let result = operation::sync::run(
            &fixture.underlying.sparse_repo_path,
            true,
            fixture.underlying.app.clone(),
        )?;
        assert!(!result.checked_out);
        assert!(!result.skipped);

        assert_eq!(result.commit_id.unwrap(), fixture.commit_id);

        Ok(())
    }

    #[test]
    fn preemptive_sync_skips_if_disabled() -> Result<()> {
        init_logging();

        let fixture = PreemptiveSyncFixture::new()?;

        fixture.repo.set_preemptive_sync_enabled(false)?;
        let result = operation::sync::run(
            &fixture.underlying.sparse_repo_path,
            true,
            fixture.underlying.app.clone(),
        )?;
        assert!(result.skipped);

        Ok(())
    }

    // Test for already being on the commit.
}
