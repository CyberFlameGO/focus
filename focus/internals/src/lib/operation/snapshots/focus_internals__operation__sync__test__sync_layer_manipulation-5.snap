---
source: focus/internals/src/lib/operation/sync.rs
assertion_line: 224
expression: "std::fs::read_to_string(&profile_path)?"
---
/*
!/*/
/3rdparty/
/focus/
/pants-internal/
/pants-support/
/tools/
/mandatory_z/

